readme.md
# Library for Probability 140

[nbviewer for Examples](https://nbviewer.jupyter.org/urls/gitlab.com/probability/prob140/raw/master/Examples.ipynb)

[Link for documentation](https://probability.gitlab.io/prob140/html/)

## DiscreteDistribution

Methods:
* `prob_event(arg)`: Finds the probability of a value or a list of values. Sums up the probabilities. Usage: `binDist.prob_event(0)` or `binDist.prob_event(np.arange(0,10))`
* `event(arg)`: Same as `prob_event(arg)` except returns a table with the probabilities rather than the sum
* `expected_value()`
* `variance()`
* `sd()`
* `plot(width=1, mask=None, **vargs)`: See Examples notebook
* `plot_event(event, **vargs)`: See Examples notebook

### FiniteDistribution

Subclass of Discrete DiscreteDistribution

Initialize with `FiniteDistribution().domain(list).probability(list)` or `FiniteDistribution().domain(list).probability_function(function)`

Examples:
```python
>>> FiniteDistribution().domain(make_array(2, 3, 4)).probability(make_array(0.25, 0.5, 0.25))
Domain | Probability
2      | 0.25
3      | 0.5
4      | 0.25

>>> FiniteDistribution().domain(make_array(1, 2, 3, 4)).probability_function(lambda x:x/10)
Domain | Probability
1      | 0.1
2      | 0.2
3      | 0.3
4      | 0.4
```

### InfiniteDistribution

Subclass of DiscreteDistribution

Initialize with `InfiniteDistribution().domain(start, end, step=1).probability_function(function)`

```python
>>> p = 0.2
>>> InfiniteDistribution().domain(1, inf).probability_function(lambda x: p*(1-p)**(x-1))
Domain           | Probability
1                | 0.2
2                | 0.16
3                | 0.128
4                | 0.1024
5                | 0.08192
6                | 0.065536
7                | 0.0524288
8                | 0.041943
9                | 0.0335544
10               | 0.0268435
... (Infinite rows omitted)
```

### JointDistribution

### Recent Changelog
* defining the domain out of order no longer messes up plots
* negative probabilities get a warnings
* renamed `p` to `prob_event`, `P` to `event`
* renamed `binWidth` to `width`
* `plot_event` and `mask` now work, but ignores width!


CHANGELOG
v0.3.0.0
--------

* Code refactor to follow PEP8

v0.2.9.0 (2017-03-19)
---------------------

* Plot_3d

v0.2.8.1 (2017-03-18)
---------------------

* Plot_continuous now accepts python functions too


v0.2.8.0 (2017-03-13)
---------------------

* Updated unconstrain to rearrange_2 and nicefy to rearrange_1

v0.2.7.1 (2017-03-11)
---------------------

* SymPy integration being finalized - added `unconstrain` and updated `declare`

v0.2.7.0 (2017-03-10)
---------------------

* sample renamed to sample_from_dist to avoid conflicts with datascience

v0.2.6.3 (2017-03-09)
---------------------

* Fixed documentation for plots
* plots removed from global

v0.2.6.2 (2017-03-09)
---------------------

* Plot_continuous works with sympy

v0.2.6.1 (2017-03-09)
---------------------

* Plot_continuous now works with any function passed in as func

v0.2.6.0 (2017-03-06)
---------------------

* Wrapper for plotting continuous functions

v0.2.5.1 (2017-03-06)
---------------------

* Beginning to add SymPy integration in *symbolic_math.py*

v0.2.5.0 (2017-02-22)
---------------------

* Added log_probability_of_path

v0.2.4.4 (2017-02-20)
---------------------

* Fixing installation issues

v0.2.4.3a (2017-02-20)
----------------------

* fixed mfpt

v0.2.4.2 (2017-02-16)
---------------------

* Fixed typo in steady_state, not sure how it happened

v0.2.4.1 (2017-02-16)
---------------------

* Documentation fix

v0.2.4.0 (2017-02-13)
---------------------

* Removed T and S from markov chains
* added .column
* states now sorted

v0.2.3.8 (2017-02-13)
---------------------

* Added get target

v0.2.3.7 (2017-02-12)
---------------------

* Deprecation error fix

v0.2.3.6 (2017-02-12)
---------------------

* Distribution now shows states with probability 0

v0.2.3.5 (2017-02-11)
---------------------

* Added show_ev for conditional distributions

v0.2.3.4 (2017-02-11)
---------------------

* state --> states

v0.2.3.3
--------
* Documentation

v0.2.3.2 (2017-02-11)
---------------------
* Changed label for empirical distribution to state
* mc.distribution accepts states

v0.2.3.1 (2017-02-11)
---------------------

* Fixed mean_first_passage_times

v0.2.3.0 (2017-02-11)
---------------------

* Renamed a ton of functions
* Implemented starting conditions

v0.2.2.0 (2017-02-11)
---------------------

* Begin wrapping of pykov

v0.2.1.3 (2017-02-08)
---------------------

* Plots uses plt.bar instead of Table.hist
* Added optional parameter edges=


v0.2.1.2 (2017-02-04)
---------------------

* Added show_ave as optional parameter

v0.2.1.1 (2017-02-04)
---------------------

* Added show_ev and show_sd as optional parameters for plot

v0.2.1.0 (2017-02-04)
---------------------

* Added sample for single variable distributions
* Added CDF for single variable distributions

v0.2.0.0 (2017-02-03)
---------------------

* Pykov

v0.1.8.1 (2017-02-01)
---------------------

* Renamed emp_dist values to proportions rather than probabilities

v0.1.8.0 (2017-01-30)
---------------------

* Added emp_dist to allow for empirical distributions


v0.1.7.6 (2017-01-19)
---------------------

* __version__ instead of version

v0.1.7.5 (2017-01-18)
---------------------

* Joint Distributions no longer give a warning if probabilities rounded to 6 decimal places = 1

v0.1.7.4 (2017-01-17)
---------------------

* Single variable distributions now check that probabilities sum to 1

v0.1.7.3 (2017-01-17)
---------------------

* Plot now adds edge border if there are fewer than 75 bins
* Plot now has an optional parameter edge that accepts a boolean
* Added marginal_dist which returns a single variable distribution

v0.1.7.2 (2017-01-17)
---------------------

* .values is now an alias for .domain

v0.1.7.1 (2017-01-17)
---------------------

* Fixed vertical axis for Plot

v0.1.7.0 (2017-01-16)
---------------------

* Removed marginal_of_X, marginal_of_Y, etc
* conditional_dist_given(given) is now conditional_dist(label, given)

v0.1.6.4 (2017-01-15)
---------------------

* Joint Distribution functions can have arbitrary number of arguments again

v0.1.6.3 (2017-01-15)
---------------------

* fixed a bug in which toJoint just renamed the x-columns rather than changing the order

v0.1.6.2 (2017-01-14)
---------------------

* toJoint now preserve original order

v0.1.6.1 (2017-01-14)
---------------------

* JointDistribution probabilities don't have to sum to 1,

v0.1.6 (2017-01-14)
-------------------

* Added probability_function for JointDistribution
* probability_function now checks number of arguments in pfunc

v0.1.5.1 (2017-01-12)
---------------------

* Added JointDistribution to the init

v0.1.5 (2017-01-12)
-------------------

* Plotting width now works with events and masks
* JointDistribution can now be used with any variable

v0.1.4.3 (2016-12-20)
---------------------

* Changed the colors for plots

v0.1.4.2
--------

* Slight modifications to plot labels

v0.1.4a
-------

* Single distribution plotting moved from the ``plot_dist`` method to the ``Plot`` function
* Multiple distribution plotting moved from the ``Plot`` function to the ``Plots`` function
* Events are now plotted by passing an argument to ``Plot``

v0.1.3
------

* Added joint distributions
* All ``FiniteDistribution`` objects changed to become ``datascience.tables.Table`` objects
* Began renaming

v0.1.2
------
Initial Release

prob140/multi_variable.py
import ast
import itertools as it
import numpy as np
import pandas as pd
import warnings

from datascience import (
    make_array,
    Table,
)


def conditional(array):
    value = array/sum(array[0:-1])
    return value


def evaluate(name):
    """
    Deletes name of RV and outputs the correct datatype, like int, float, or
    string.

    Parameters
    ----------
    name : String
        In the form 'rv=123123124'.

    Returns
    -------
    String, int, float
    """
    try:
        index = name.index('=')
    except Exception:
        index = name.index(' ')
    try:
        return ast.literal_eval(name[index + 1:])
    except Exception:
        return name[index + 1:]


class JointDistribution(pd.DataFrame):

    def marginal(self, label):
        """
        Returns the marginal distribution of label

        Parameters
        ----------
        label : String
            The label of the variable of which we want to find the marginal
            distribution.

        Returns
        -------
        JointDistribution Table

        Examples
        --------
        >>> dist2 = Table().values('Coin1',['H','T'],'Coin2', ['H','T']).probability(np.array([0.24, 0.36, 0.16, 0.24])).to_joint()
        >>> dist2.marginal('Coin1')
                                Coin1=H  Coin1=T
        Coin2=T                    0.36     0.24
        Coin2=H                    0.24     0.16
        Sum: Marginal of Coin1     0.60     0.40
        >>> dist2.marginal('Coin2')
                 Coin1=H  Coin1=T  Sum: Marginal of Coin2
        Coin2=T     0.36     0.24                     0.6
        Coin2=H     0.24     0.16                     0.4
        """
        copy = JointDistribution(self, copy=True)
        if label == self._X_column_label:
            key = 'Sum: Marginal of {0}'.format(self._X_column_label)
            copy.loc[key] = copy.sum(axis=0)
        elif label == self._Y_column_label:
            key = 'Sum: Marginal of {0}'.format(self._Y_column_label)
            copy[key] = copy.sum(axis=1)
        else:
            raise AssertionError(
                'Label does not correspond with existing variable name')
        return copy

    def marginal_dist(self, label):
        """
        Finds the marginal marginal distribution of label, returns as a single
        variable distribution.

        Parameters
        ----------
        label
            The label of the variable of which we want to find the marginal
            distribution.

        Returns
        -------
        Table
            Single variable distribution of label.
        """
        marginal = self.marginal(label).as_matrix()
        if label == self._X_column_label:
            x_labels = list(self)
            prob = marginal[-1, :]
            domain = [evaluate(lab) for lab in x_labels]
            return Table().values(domain).probability(prob)
        else:
            y_labels = list(self.index)
            prob = marginal[:, -1]
            domain = [evaluate(lab) for lab in y_labels]
            return Table().values(domain).probability(prob)

    def both_marginals(self):
        """
        Finds the marginal distribution of both variables.

        Returns
        -------
        JointDistribution Table.

        Examples
        --------
        >>> dist1 = Table().values([0,1],[2,3]).probability([0.1, 0.2, 0.3, 0.4]).to_joint()
        >>> dist1.both_marginals()
                            X=0  X=1  Sum: Marginal of Y
        Y=3                 0.2  0.4                 0.6
        Y=2                 0.1  0.3                 0.4
        Sum: Marginal of X  0.3  0.7                 1.0
        """
        copy = JointDistribution(self, copy=True)
        key_y = 'Sum: Marginal of {0}'.format(self._Y_column_label)
        key_x = 'Sum: Marginal of {0}'.format(self._X_column_label)
        copy[key_y] = copy.sum(axis=1)
        copy.loc[key_x] = copy.sum(axis=0)
        return copy

    def conditional_dist(self, label, given='', show_ev=False):
        """
        Given the random variable label, finds the conditional distribution of
        the other variable.

        Parameters
        ----------
        label : String
            Variable given.

        Returns
        -------
        JointDistribution Table

        Examples
        --------
        >>> coins = Table().values('Coin1',['H','T'],'Coin2', ['H','T']).probability(np.array([0.24, 0.36, 0.16,0.24])).to_joint()
        >>> coins.conditional_dist('Coin1','Coin2')
                                  Coin1=H  Coin1=T  Sum
        Dist. of Coin1 | Coin2=H      0.6      0.4  1.0
        Dist. of Coin1 | Coin2=T      0.6      0.4  1.0
        Marginal of Coin1             0.6      0.4  1.0
        >>> coins.conditional_dist('Coin2','Coin1')
                 Dist. of Coin2 | Coin1=H  Dist. of Coin2 | Coin1=T  Marginal of Coin2
        Coin2=H                       0.4                       0.4                0.4
        Coin2=T                       0.6                       0.6                0.6
        Sum                           1.0                       1.0                1.0
        """
        if label == self._Y_column_label:
            both = self.both_marginals()
            new = np.append(both.index[0:-1], 'Sum')
            y = both.apply(conditional, axis=0).set_index(new)
            matrix = y.as_matrix()[:-1, :]
            y_labels = list(self.index)
            domain = np.array([evaluate(lab) for lab in y_labels])
            exp_values = [sum(matrix[:, i] * domain)
                          for i in range(len(matrix[0]))]
            column_names = y.columns

            new = make_array()
            for i in np.arange(len(column_names) - 1):
                new_name = 'Dist. of {0} | '.format(self._Y_column_label)
                new_name += column_names[i]
                new = np.append(new, new_name)
            new = np.append(new, 'Marginal of {0}'.format(self._Y_column_label))
            y.columns = new
            if show_ev:
                y.loc['EV'] = exp_values
            return y

        elif label == self._X_column_label:
            both = self.both_marginals()

            x = both.apply(conditional, axis=1).rename(columns={
                'Sum: Marginal of {0}'.format(self._Y_column_label): 'Sum'})

            matrix = x.as_matrix()[:, :-1]
            x_labels = list(self)
            domain = np.array([evaluate(lab) for lab in x_labels])
            exp_values = [sum(matrix[i] * domain) for i in range(len(matrix))]
            indices = both.index
            new = make_array()
            for i in np.arange(len(indices) - 1):
                new_name = 'Dist. of {0} | '.format(self._X_column_label)
                new_name += indices[i]
                new = np.append(new, new_name)
            new = np.append(new, 'Marginal of {0}'.format(self._X_column_label))
            new_df = x.set_index(new)

            if show_ev:
                new_df['EV'] = exp_values

            return new_df
        else:
            raise AssertionError(
                'Label does not correspond with existing variable name')


def multi_domain(table, *args):

    if isinstance(args[0], str):
        assert len(args) % 2 == 0, 'Must alternate between name and values'
        var_names = [args[2 * i] for i in range(len(args) // 2)]
        values = [args[2 * i + 1] for i in range(len(args) // 2)]
        var_values = list(zip(*it.product(*values)))
    else:
        var_names = [chr(ord('X') + i) for i in range(len(args))]
        var_values = list(zip(*it.product(*args)))

    new_table = table.copy()
    for column_name, column_value in reversed(list(zip(var_names, var_values))):
        new_table = new_table.with_column(column_name, column_value)
        new_table.move_to_start(column_name)

    return new_table


def to_joint(table, X_column_label=None, Y_column_label=None,
             probability_column_label=None, reverse=True):
    """
    Converts a table of probabilities associated with two variables into a
    JointDistribution object

    Parameters
    ----------
    table : Table
        You can either call pass in a Table directly or call the toJoint()
        method of that Table. See examples.
    X_column_label (optional) : str
        Label for the first variable. Defaults to the same label as that of
        first variable of Table.
    Y_column_label (optional) : str
        Label for the second variable. Defaults to the same label as that of
        second variable of Table.
    probability_column_label (optional) : str
        Label for probabilities.
    reverse (optional) : bool
        If True, the vertical values will be reversed.

    Returns
    -------
    JointDistribution
        A JointDistribution object.

    Examples
    --------
    >>> dist1 = Table().values([0,1],[2,3])
    >>> dist1['Probability'] = make_array(0.1, 0.2, 0.3, 0.4)
    >>> dist1.to_joint()
         X=0  X=1
    Y=3  0.2  0.4
    Y=2  0.1  0.3
    >>> dist2 = Table().values('Coin1',['H','T'], 'Coin2', ['H','T'])
    >>> dist2['Probability'] = np.array([0.4*0.6, 0.6*0.6, 0.4*0.4, 0.6*0.4])
    >>> dist2.toJoint()
             Coin1=H  Coin1=T
    Coin2=T     0.36     0.24
    Coin2=H     0.24     0.16

    """
    assert table.num_columns >= 3, (
        'You must have columns for your X variable, for your Y variable, and '
        'for your probabilities')
    if X_column_label is None:
        X_column_label = table.labels[0]
    if Y_column_label is None:
        Y_column_label = table.labels[1]
    if probability_column_label is None:
        probability_column_label = table.labels[table.num_columns-1]

    total = sum(table[probability_column_label])

    if round(total, 6) != 1:
        warnings.warn('Your probabilities sum to {0}'.format(total))

    x_possibilities = sorted(set(table[X_column_label]))
    y_possibilities = sorted(set(table[Y_column_label]), reverse=reverse)

    xInd = table.column_index(X_column_label)
    yInd = table.column_index(Y_column_label)
    pInd = table.column_index(probability_column_label)

    data = {poss: [0]*len(y_possibilities) for poss in x_possibilities}

    for row in table.rows:
        data[row[xInd]][y_possibilities.index(row[yInd])] += row[pInd]

    x_order = ['{}={}'.format(X_column_label, poss) for poss in x_possibilities]

    realData = {'{}={}'.format(X_column_label, str(poss)): value
                for poss, value in data.items()}
    index = ['{}={}'.format(Y_column_label, poss) for poss in y_possibilities]

    # Reverting order back to original
    df = pd.DataFrame(realData, index=index)
    joint_dist = JointDistribution(df[x_order], index=index)

    joint_dist.reindex(index)

    joint_dist._X_column_label = X_column_label
    joint_dist._Y_column_label = Y_column_label

    return joint_dist
